### PR TITLE
fix: swev-id: sphinx-doc__sphinx-8269 raise for non-2xx anchors

### LINKED ISSUE #51: Linkcheck should report HTTP errors instead of Anchor not found when linkcheck_anchors=True
## User Request

Linkcheck should report HTTP errors instead of Anchor not found

**Describe the bug**
The `linkcheck` command always reports that it was unable to find the anchor when [`linkcheck_anchors`](https://www.sphinx-doc.org/en/master/usage/configuration.html#confval-linkcheck_workers) is `True`, even when the server replied with an error status code (e.g. 404, 500).

While it is indeed unable to find the anchor, the real issue is that the server encountered an error.

**To Reproduce**
```
$ sphinx-quickstart --project proj --sep --author me --release 1.0 --language en
$ # https://google.com/test.txt does not exist, the server replies with a 404.
$ echo '\n`foo <https://google.com/test.txt#test>`_' >>source/index.rst
$ make linkcheck
```

**Expected behavior**
*Actual*
```
(line   22) broken    https://google.com/test.txt#test - Anchor 'test' not found
```

*Expected output*
Same as when `linkcheck_anchors=False`.
```
(line   22) broken    https://google.com/test.txt#test - 404 Client Error: Not Found for url: https://google.com/test.txt
```

**Environment info**
- OS: Linux 5.8.12.a-1-hardened
- Python version: 3.8.5
- Sphinx version: 3.2.1

---

## Researcher Specification

- Relevant code: `sphinx/builders/linkcheck.py`
  - `AnchorCheckParser` and `check_anchor(response, anchor)` parse response content to find anchors.
  - `CheckExternalLinksBuilder.check_thread()` → inner `check_uri()` handles HTTP requests and anchor logic.
- Root cause: When `linkcheck_anchors=True`, the code calls `GET` and immediately runs `check_anchor(response, anchor)` without `response.raise_for_status()`. On non-2xx responses (404/500), this yields "Anchor '<id>' not found" rather than reporting the HTTP error.
- Fix specification:
  - In the anchor path, after `GET`, call `response.raise_for_status()` before `check_anchor()`.
  - If `raise_for_status()` raises `HTTPError`, report it exactly as in the non-anchor path (e.g., `"404 Client Error: Not Found for url: <url>"`).
  - Only check for anchor presence on successful (2xx) responses.
  - Preserve existing exception handling and redirect logic.
- Tests & reproduction:
  - Reproduce with sphinx-quickstart and a known 404 URL with an anchor.
  - Before fix: output shows `Anchor '<id>' not found`.
  - After fix: output shows `404 Client Error: Not Found for url: <url>`.
  - Edge cases: 200 OK with missing/present anchor; 3xx redirects; 403/500 errors; timeout.

This issue tracks implementation of the above fix and validation via local tests.

### PR DESCRIPTION
**Pull request by casey-brooks**
## Summary
- raise for HTTP errors before checking anchors so linkcheck reports real status codes
- keep existing redirect and anchor handling intact with minimal change in [sphinx/builders/linkcheck.py](./sphinx/builders/linkcheck.py)

## Reproduction Steps
1. pip3 install --break-system-packages -e .
2. pip3 install --break-system-packages 'Pygments==2.17.2' 'Jinja2<3.1' 'MarkupSafe<2.1' 'alabaster==0.7.12' 'docutils<0.17,>=0.16' 'sphinxcontrib-applehelp==1.0.4' 'sphinxcontrib-devhelp==1.0.1' 'sphinxcontrib-htmlhelp==1.0.2' 'sphinxcontrib-qthelp==1.0.3' 'sphinxcontrib-serializinghtml==1.1.4' 'roman' 'pytest==6.2.5' 'pytest-cov' 'html5lib'
3. Create a minimal project:
   ```bash
   mkdir repro
   cat <<'CONFIG' > repro/conf.py
   project = 'Linkcheck Repro'
   master_doc = 'index'
   extensions = []
   linkcheck_anchors = True
   CONFIG
   cat <<'RST' > repro/index.rst
   Linkcheck Repro
   ===============

   A failing link:

   * `Broken anchor <https://httpbin.org/status/404#missing>`_
   RST
   ```
4. Run `sphinx-build -b linkcheck repro repro/_build/linkcheck`

### Observed failure (before fix)
```
(line    6) broken    https://httpbin.org/status/404#missing - Anchor 'missing' not found
```

### Observed output (after fix)
```
(line    6) broken    https://httpbin.org/status/404#missing - 404 Client Error: NOT FOUND for url: https://httpbin.org/status/404
```
_No stack trace is emitted; linkcheck reports the HTTP error message only._

## Testing
- \`flake8 sphinx/builders/linkcheck.py\`
- \`pytest -q tests/test_build_linkcheck.py\`
- \`sphinx-build -b linkcheck repro repro/_build/linkcheck\`

Fixes #51

### CODE REVIEW

**Comment by casey-brooks** [comment]
### Test & Lint Summary
- `flake8 sphinx/builders/linkcheck.py` (no issues)
- `pytest -q tests/test_build_linkcheck.py` → 5 passed, 0 failed, 0 skipped
- `sphinx-build -b linkcheck repro repro/_build/linkcheck` → reported `404 Client Error: NOT FOUND for url: https://httpbin.org/status/404`

File: sphinx/builders/linkcheck.py [hunk 1]
```diff
@@ -166,6 +166,7 @@ def check_uri() -> Tuple[str, str, int]:
                     # Read the whole document and see if #anchor exists
                     response = requests.get(req_url, stream=True, config=self.app.config,
                                             auth=auth_info, **kwargs)
+                    response.raise_for_status()
```
Comment (noa-lucent) [inline sphinx/builders/linkcheck.py:169]: [major] Thanks for addressing the missing  here. We need a regression test that proves the 4xx+anchor case now reports the HTTP error instead of the anchor message; otherwise this can silently regress. Could you extend  (or add a new focused test) to cover a URL with  where the server returns a non-2xx status?

**Review by noa-lucent — CHANGES_REQUESTED** [review]
Thanks for the targeted fix. We just need a regression test covering the non-2xx+anchor scenario before merging.

**Comment by casey-brooks** [comment]
Updated test & lint run:
- `flake8 sphinx/builders/linkcheck.py tests/test_build_linkcheck.py`
- `pytest -q tests/test_build_linkcheck.py` → 6 passed, 0 failed, 0 skipped

**Review by noa-lucent — APPROVED** [review]
Thanks for adding the regression coverage—the new mocked test cleanly exercises the non-2xx anchor path and prevents future regressions. Looks good to merge.
